### PR TITLE
T gen upscope

### PR DESCRIPTION
**Pull request by pawelsag**
We appreciate your contributing to Verilator.  If this is your first commit, please add your name to docs/CONTRIBUTORS, and read our contributing guidelines in docs/CONTRIBUTING.adoc.

### CODE REVIEW

File: src/V3EmitC.cpp [hunk 1]
```diff
@@ -1385,8 +1385,19 @@ class EmitCStmts VL_NOT_FINAL : public EmitCBaseVisitor {
         AstNodeDType* dtypep = nodep->varp()->dtypep();
         bool useDly = useDelayedValue(nodep);
         if (useDly) puts("verilated_nba_ctrl.get_scheduled(&");
+        if(nodep->abovep() != nullptr)
+            if(nodep->isWide() and nodep->abovep()->type() == AstType::atIf)
```
Comment (kbieganski) [inline src/V3EmitC.cpp:1389]: Please use the `VN_IS` macro for checking AST types (`VN_IS(nodep, If)`)

File: src/V3Ast.cpp [hunk 2]
```diff
@@ -93,7 +93,10 @@ AstNode* AstNode::abovep() const {
     // m_headtailp only valid at beginning or end of list
     // Avoid supporting at other locations as would require walking
     // list which is likely to cause performance issues.
-    UASSERT_OBJ(!m_nextp || firstAbovep(), this, "abovep() not allowed when in midlist");
```
Comment (kbieganski) [inline src/V3Ast.cpp:96]: We should probably avoid modifying core AST functions unless strictly necessary. Could we not do this differently?

File: src/V3EmitC.cpp [hunk 3]
```diff
@@ -2438,9 +2449,9 @@ void EmitCStmts::emitOpName(AstNode* nodep, const string& format, AstNode* lhsp,
                     needComma = true;
                     break;
                 case 'W':
-                    if (lhsp->isWide()) {
```
Comment (kbieganski) [inline src/V3EmitC.cpp:2441]: Does this not break other operators which should have `lhsp` here?

File: src/V3GraphAcyc.cpp [hunk 4]
```diff
@@ -345,7 +345,7 @@ void GraphAcyc::simplifyOut(GraphAcycVertex* avertexp) {
                 nextp = inEdgep->inNextp();
                 V3GraphVertex* inVertexp = inEdgep->fromp();
                 if (inVertexp == avertexp) {
-                    // cutting of the cycles, exactly in the first node causing a cycle
+                    // TODO psagan cutting of the cycles, exactly in the first node causing a cycle
```
Comment (kbieganski) [inline src/V3GraphAcyc.cpp:348]: What is there TODO?

File: src/V3GraphAlg.cpp [hunk 5]
```diff
@@ -280,8 +280,9 @@ class GraphAlgRank final : GraphAlg<> {
         // If larger rank is found, assign it and loop back through
         // If we hit a back node make a list of all loops
         if (vertexp->user() == 1) {
-            m_graphp->reportLoops(m_edgeFuncp, vertexp);
-            m_graphp->loopsMessageCb(vertexp);
+            vertexp->unlinkDelete(m_graphp);
+           // m_graphp->reportLoops(m_edgeFuncp, vertexp);
```
Comment (kbieganski) [inline src/V3GraphAlg.cpp:284]: Should we uncomment this code at some point?

Comment (pawelsag) [reply src/V3GraphAlg.cpp:284]: Yes, I'm currently changing this implementation. I found that changing this so simply wouldn't work
Cause with this implementation we wouldn't be able to distinguish from real cycles.

File: src/V3Ast.cpp (hunk 2 above)
Comment (pawelsag) [reply src/V3Ast.cpp]: I agree, I restored old implementation

File: src/V3EmitC.cpp (hunk 1 above)
Comment (pawelsag) [reply src/V3EmitC.cpp]: thanks for this remark

File: src/V3EmitC.cpp (hunk 3 above)
Comment (pawelsag) [reply src/V3EmitC.cpp:2435]: It should work, cause before we are using this pointer we are choosing the right node

**Pull request merged**